### PR TITLE
Improve performance

### PR DESCRIPTION
**Pull request by garrettheel**
This introduces a number of performance improvements to help make this plugin feasible for larger documentation sets.

Performance:

 * Use the faster `lxml` parser for BeautifulSoup
 * Use a `SoupStrainer` to only consider elements that should represent links  
 * Compile regular expressions once and store on the module, rather than re-compiling repeatedly
 * Move `lru_cache` to a dedicated staticmethod to fetch external URLs. This allows external URLs to be cached between pages. Note that serializing the `BeautifulSoup` arg for use with the cache was very expensive before, so this also avoids that.
 * Use a `requests.Session` for HTTP requests
 * Make `HEAD` requests instead of `GET`. Note that some sites may not handle this correctly, but generally this seemed to work fine 

Features:

 * Adds support for `<a id="blah"></a>` links in markdown pages, where people may want to support custom links in addition to the slugified headings.

Other:

 * Only prints errors to stdout when `raise_error_excludes` does not match. This reduces unnecessary noise
 * Changes the output format for errors to `invalid url - {clean_url} [{url_status}] [{page.file.src_path}]` to provide more information

### CODE REVIEW

**Comment by manuzhang** [comment]
@johnthagen please take a look as well.

**Comment by garrettheel** [comment]
@johnthagen let me know if you have any questions/concerns! I was looking to submit a few additional config options after this lands too

**Comment by johnthagen** [comment]
@garrettheel Sorry for the delay. I haven't had time to look at this yet. Hopefully can this week, otherwise if @manuzhang thinks it's good please don't feel the need to have to wait for me.

File: htmlproofer/plugin.py
Comment (johnthagen) [inline htmlproofer/plugin.py:18]: Any reason not to simply make `lxml` a hard dependency?

Comment (garrettheel) [reply htmlproofer/plugin.py:18]: Mostly that it is a heavier dependency so I was being conservative. That said, lxml has wheels available for most platforms so I don't think it's unreasonable to just depend on it

Comment (garrettheel) [reply htmlproofer/plugin.py]: Just pushed a commit to do this

File: README.md
Comment (johnthagen) [inline README.md:74]: Let's change this now that lxml is a hard dependency.

Comment (garrettheel) [reply README.md]: good call, done

File: htmlproofer/plugin.py
Comment (johnthagen) [inline htmlproofer/plugin.py:66]: What is the purpose of using an `@lru_cache` with a size of `1`?

Comment (garrettheel) [reply htmlproofer/plugin.py:66]: The default is 128, so this is to be explicit that only 1 would ever be used

File: htmlproofer/plugin.py
Comment (johnthagen) [inline htmlproofer/plugin.py:162]: Is this line indented properly? It looks like it needs two more spaces

File: htmlproofer/plugin.py
Comment (johnthagen) [reply htmlproofer/plugin.py:66]: Ah, this makes sense now. Since there is no constructor in this setup this seems like an appropriate thing to do.

**Review by johnthagen — APPROVED** [review]

File: htmlproofer/plugin.py
Comment (garrettheel) [reply htmlproofer/plugin.py]: Actually I think you're right that a constructor makes more sense. I was overlooking the fact that the same plugin instance is used for the lifetime of the build

**Comment by garrettheel** [comment]
@johnthagen this should be ready to go from my end

one thing I noticed was that `docs/index.md` duplicates most of `README.md` and will become out of date. would it make sense put something more generic in there as a test, rather than the actual docs? or move it to a directory name that indicates that it's used only for testing

**Comment by johnthagen** [comment]
@manuzhang set up that originally, but I do agree that it would make sense to have the test's be isolated from the real docs. That would allow the test project to focus specifically on the items we wish to test.

**Comment by garrettheel** [comment]
> @manuzhang set up that originally, but I do agree that it would make sense to have the test's be isolated from the real docs. That would allow the test project to focus specifically on the items we wish to test.

Cool, I'm happy to change that here if you like. Otherwise we can handle that in a follow-up

**Comment by johnthagen** [comment]
@garrettheel Thank you for being willing to help out. Let's make a follow on issue for that so we can review it in pieces.

**Review by johnthagen — APPROVED** [review]

**Comment by garrettheel** [comment]
@johnthagen no worries! I don't have write access so you'll have to do merge for me when you're ready :)

**Comment by johnthagen** [comment]
@garrettheel I don't have write access either, I just contributed a few PRs. @manuzhang Will need to do the final review/merge.

**Comment by garrettheel** [comment]
Whoops, my bad. @manuzhang when you get a chance please take another look :)

**Review by manuzhang — APPROVED** [review]

**Comment by manuzhang** [comment]
@garrettheel @johnthagen thanks guys. Merged.

**Pull request merged**
